k2/python/csrc/torch/index_add.cu

```diff
@@ -52,68 +57,149 @@ void IndexAdd(torch::Tensor index, torch::Tensor value, torch::Tensor *in_out) {
 
   if (index.is_contiguous() && value.is_contiguous() &&
       in_out->is_contiguous()) {
-    if (context->GetDeviceType() == kCpu) {
-      for (int32_t i = 0; i != index_numel; ++i) {
-        int32_t in_out_i = index_data[i];
-        if (in_out_i == -1) continue;
+    K2_EVAL(
+        context, index_numel, lambda_contiguous, (int32_t i)->void {
+          int32_t in_out_i = index_data[i];
+          if (in_out_i == -1) return;
+
+          K2_DCHECK_GE(in_out_i, 0);
+          K2_DCHECK_LT(in_out_i, in_out_numel);
+
+#ifdef __CUDA_ARCH__
+          atomicAdd(in_out_data + in_out_i, value_data[i]);
+#else
+          // for host thread, we assume single thread at present
+          in_out_data[in_out_i] += value_data[i];
+#endif
+        });
+    return;
+  }
 
-        K2_DCHECK_GE(in_out_i, 0);
-        K2_DCHECK_LT(in_out_i, in_out_numel);
+  // for non-contiguous tensors
+  int64_t index_stride = index.strides()[0];
+  int64_t value_stride = value.strides()[0];
+  int64_t in_out_stride = in_out->strides()[0];
 
-        // for host thread, we assume single thread at present
-        in_out_data[in_out_i] += value_data[i];
-      }
-    } else {
-      // for cuda
-      auto lambda_contiguous = [=] __device__(int32_t i) -> void {
-        int32_t in_out_i = index_data[i];
+  K2_EVAL(
+      context, index_numel, lambda_noncontiguous, (int32_t i)->void {
+        int32_t in_out_i = index_data[i * index_stride];
         if (in_out_i == -1) return;
 
         K2_DCHECK_GE(in_out_i, 0);
         K2_DCHECK_LT(in_out_i, in_out_numel);
 
-        atomicAdd(in_out_data + in_out_i, value_data[i]);
-      };
-      EvalDevice(context, index_numel, lambda_contiguous);
-    }
+#ifdef __CUDA_ARCH__
+        atomicAdd(in_out_data + in_out_i * in_out_stride,
+                  value_data[i * value_stride]);
+#else
+      // for host thread, we assume single thread at present
+      in_out_data[in_out_i * in_out_stride] += value_data[i * value_stride];
+#endif
+      });
+}
+
+static void IndexAdd2D(torch::Tensor index, torch::Tensor value,
```

Comment:
There is no documentation here!!
You should be writing the documentation first!

Reply:
I can't review this without documentation- how can I tell if it is doing what is intended when that is not stated?

Reply:
Removed.